File: instabiz/instabiz/page/ib_business_pulse/ib_business_pulse.js
frappe.pages["ib-business-pulse"].on_page_load = function (wrapper) {
	frappe.ui.make_app_page({
		parent: wrapper,
		title: "Business Pulse",
		single_column: true,
	});
	wrapper._ib_pulse = new IBBusinessPulse(wrapper);
};

frappe.pages["ib-business-pulse"].on_page_show = function (wrapper) {
	if (!wrapper._ib_pulse) return;
	wrapper._ib_pulse.refresh();
	wrapper._ib_pulse._start_auto();
};

frappe.pages["ib-business-pulse"].on_page_hide = function (wrapper) {
	if (wrapper._ib_pulse) wrapper._ib_pulse._stop_auto();
};

// ─────────────────────────────────────────────────────────────────────────────
// Each domain card shows real counts/amounts only — no synthetic 0-100 score.
// Every metric routes to the actual filtered record list it was counted from;
// the card footer routes to the domain's own dashboard page.

class IBBusinessPulse {
	constructor(wrapper) {
		this.wrapper = wrapper;
		this.page = wrapper.page;
		this._data = null;
		this._trend_chart = null;
		this._auto_timer = null;
		this._inject_styles();
		this._build_layout();
		this._bind_toolbar();
		// refresh + timer started by on_page_show to avoid double call on first load
	}

	_inject_styles() {
		if (document.getElementById("ib-bp-styles")) return;
		const s = document.createElement("style");
		s.id = "ib-bp-styles";
		s.textContent = `
.ib-bp-wrap { padding: 16px; max-width: 1400px; }
.ib-bp-domain-grid { display: grid; grid-template-columns: repeat(3, 1fr); gap: 12px; margin-bottom: 20px; }
.ib-bp-domain { background: var(--card-bg); border: 1px solid var(--border-color);
  border-radius: 8px; overflow: hidden; }
.ib-bp-domain-top { display: flex; align-items: center; gap: 8px; padding: 12px 14px 10px; }
.ib-bp-domain-icon { display: inline-flex; }
.ib-bp-domain-name { font-size: 12.5px; font-weight: 600; color: var(--heading-color); }
.ib-bp-metric-row { display: flex; align-items: center; justify-content: space-between;
  padding: 8px 14px; border-top: 1px solid var(--border-color); cursor: pointer; transition: background .12s; }
.ib-bp-metric-row:hover { background: var(--subtle-fg, rgba(0,0,0,.03)); }
.ib-bp-metric-lbl { font-size: 11.5px; color: var(--text-muted); }
.ib-bp-metric-val { font-size: 14px; font-weight: 700; color: var(--heading-color); }
.ib-bp-metric-badge { font-size: 10px; font-weight: 600; margin-left: 6px; }
.ib-bp-metric-badge.up { color: #10b981; }
.ib-bp-metric-badge.down { color: #ef4444; }
.ib-bp-domain-footer { display: block; padding: 7px 14px; font-size: 10.5px; font-weight: 600;
  color: var(--ib-primary, #d97757); text-align: right; border-top: 1px solid var(--border-color);
  cursor: pointer; text-transform: uppercase; letter-spacing: .04em; }
.ib-bp-domain-footer:hover { text-decoration: underline; }
.ib-bp-card { background: var(--card-bg); border: 1px solid var(--border-color); border-radius: 8px; padding: 16px; margin-bottom: 16px; }
.ib-bp-card-title { font-size: 12px; font-weight: 600; color: var(--text-muted);
  text-transform: uppercase; letter-spacing: .5px; margin-bottom: 14px; }
.ib-bp-trend-wrap { height: 180px; }
.ib-bp-ts { font-size: 11px; color: var(--text-muted); text-align: right; margin-bottom: 12px; }
@media(max-width:900px){ .ib-bp-domain-grid{grid-template-columns:1fr 1fr;} }
@media(max-width:540px){ .ib-bp-domain-grid{grid-template-columns:1fr;} }
		`;
		document.head.appendChild(s);
	}

	_build_layout() {
		const $pc = $(this.wrapper).find(".page-content");
		this.$wrap = $(`<div class="ib-bp-wrap"></div>`).appendTo($pc);
		this.$wrap.html(`
			<div class="ib-bp-ts" id="ib-bp-ts">Loading…</div>
			<div class="ib-bp-domain-grid" id="ib-bp-domains"></div>
			<div class="ib-bp-card">
				<div class="ib-bp-card-title">14-Day Revenue Trend</div>
				<div class="ib-bp-trend-wrap" id="ib-bp-trend"></div>
			</div>
		`);
	}

	_bind_toolbar() {
		this.page.add_button(__("Dashboard"), () => frappe.set_route("ib-main-dashboard"));
		this.page.add_button(__("Analytics Hub"), () => frappe.set_route("ib-analytics-hub"));
		this.page.add_inner_button(__("Refresh"), () => this.refresh());
	}

	refresh() {
		this.$wrap.find("#ib-bp-ts").text("Loading…");
		frappe.call({
			method: "instabiz.instabiz.page.ib_business_pulse.ib_business_pulse.get_pulse_data",
			callback: (r) => {
				if (r.message) {
					this._data = r.message;
					this._render(r.message);
					this.$wrap.find("#ib-bp-ts").text("Updated " + frappe.datetime.now_time() + " · Auto-refresh every 2 min");
				}
			},
			error: () => this.$wrap.find("#ib-bp-ts").text("Error loading data"),
		});
	}

	_start_auto() {
		if (this._auto_timer) return;
		this._auto_timer = setInterval(() => this.refresh(), 120000);
	}

	_stop_auto() {
		if (this._auto_timer) { clearInterval(this._auto_timer); this._auto_timer = null; }
	}

	_fmt(v) { return "₹" + Number(v || 0).toLocaleString("en-IN", { maximumFractionDigits: 0 }); }

	_render(d) {
		this._render_domains(d);
		this._render_trend(d.trend_14 || []);
	}

	// Route filters below mirror the same date_field/doctype the backend just
	// aggregated with (billing_mode toggle — Sales Order in dev, Sales Invoice
	// in prod) so a click always lands on the exact rows that were counted.
	_domains(d) {
		const doctype = d.sales_doctype;
		const date_field = d.date_field;
		const between = [d.month_start, d.today];
		const status_filter = d.dev_mode ? { status: ["!=", "Cancelled"] } : { is_return: 0 };

		return [
			{
				key: "Revenue", icon: "lucide:indian-rupee", route: "ib-main-dashboard",
				metrics: [
					{
						label: "Revenue MTD", value: this._fmt(d.rev_mtd),
						badge: d.rev_change_pct == null ? "" :
							`<span class="ib-bp-metric-badge ${d.rev_change_pct >= 0 ? "up" : "down"}">${d.rev_change_pct >= 0 ? "▲" : "▼"} ${Math.abs(d.rev_change_pct)}% vs last mo</span>`,
						route: () => frappe.set_route("List", doctype, Object.assign({ docstatus: 1, [`${date_field},Between`]: between }, status_filter)),
					},
					{
						label: "Collection Rate", value: d.collection_rate + "%",
						route: () => frappe.set_route("query-report", "IB Collections Report"),
					},
					{
						label: "Outstanding AR", value: this._fmt(d.ar),
						route: () => frappe.set_route("query-report", "IB AR Aging"),
					},
				],
			},
			{
				key: "Sales", icon: "lucide:file-text", route: "ib-customer-board",
				metrics: [
					{
						label: "Open Leads", value: d.open_leads,
						route: () => frappe.set_route("List", "Lead", { status: ["not in", ["Converted", "Do Not Contact"]] }),
					},
					{
						label: "Open Quotations", value: d.open_quotes,
						route: () => frappe.set_route("List", "Quotation", { docstatus: 1, status: ["not in", ["Ordered", "Lost", "Cancelled", "Expired"]] }),
					},
				],
			},
			{
				key: "Inventory", icon: "lucide:package", route: "ib-stock-dashboard",
				metrics: [
					{ label: "Items In Stock", value: d.total_items, route: () => frappe.set_route("ib-stock-dashboard") },
					{ label: "Low / Reorder Stock", value: d.low_stock, route: () => frappe.set_route("ib-stock-dashboard") },
				],
			},
			{
				key: "Procurement", icon: "lucide:shopping-cart", route: "ib-procurement-dashboard",
				metrics: [
					{
						label: "Open Purchase Orders", value: d.open_po,
						route: () => frappe.set_route("List", "Purchase Order", { docstatus: 1, status: ["not in", ["Completed", "Cancelled", "Closed"]] }),
					},
				],
			},
			{
				key: "HR", icon: "lucide:users", route: "ib-hrms-dashboard",
				metrics: [
					{ label: "Active Employees", value: d.total_emp, route: () => frappe.set_route("List", "Employee", { status: "Active" }) },
					{
						label: "Present Today", value: d.present_today,
						route: () => frappe.set_route("List", "Attendance", { attendance_date: d.today, status: "Present", docstatus: 1 }),
					},
				],
			},
			{
				key: "Production", icon: "lucide:factory", route: "ib-production-dashboard",
				metrics: [
					{
						label: "Active Work Orders", value: d.wo_active,
						route: () => frappe.set_route("List", "IB Work Order", { status: ["in", ["Pending", "In Progress", "On Hold"]] }),
					},
					{
						label: "Completed This Month", value: d.wo_completed,
						route: () => frappe.set_route("List", "IB Work Order", { status: "Completed", "completed_at,>=": d.month_start }),
					},
				],
			},
		];
	}

	_render_domains(d) {
		const domains = this._domains(d);
		const html = domains.map((dom, di) => `
			<div class="ib-bp-domain">
				<div class="ib-bp-domain-top">
					<iconify-icon icon="${dom.icon}" width="18" height="18" class="ib-bp-domain-icon"></iconify-icon>
					<span class="ib-bp-domain-name">${dom.key}</span>
				</div>
				${dom.metrics.map((m, mi) => `
					<div class="ib-bp-metric-row" data-domain="${di}" data-metric="${mi}">
						<span class="ib-bp-metric-lbl">${m.label}</span>
						<span><span class="ib-bp-metric-val">${m.value}</span>${m.badge || ""}</span>
					</div>
				`).join("")}
				<div class="ib-bp-domain-footer" data-route="${dom.route}">Open ${dom.key} Dashboard →</div>
			</div>
		`).join("");

		const $el = this.$wrap.find("#ib-bp-domains").html(html);
		$el.find(".ib-bp-metric-row").on("click", (e) => {
			const $row = $(e.currentTarget);
			const dom = domains[$row.data("domain")];
			const metric = dom.metrics[$row.data("metric")];
			metric.route();
		});
		$el.find(".ib-bp-domain-footer").on("click", (e) => {
			frappe.set_route($(e.currentTarget).data("route"));
		});
	}

	_render_trend(trend) {
		const $el = this.$wrap.find("#ib-bp-trend")[0];
		if (!$el) return;
		if (!trend.length) {
			$($el).html(`<div style="padding:30px;text-align:center;color:var(--text-muted);font-size:12px">No trend data</div>`);
			return;
		}
		if (this._trend_chart) { this._trend_chart.destroy && this._trend_chart.destroy(); this._trend_chart = null; }
		$($el).empty();
		this._trend_chart = new frappe.Chart($el, {
			type: "line",
			data: {
				labels: trend.map(r => r.label),
				datasets: [{ name: "Revenue", values: trend.map(r => parseFloat(r.amount || 0)) }],
			},
			colors: ["#d97757"],
			height: 165,
			lineOptions: { regionFill: 1, hideDots: 1, spline: 1 },
			axisOptions: { xIsSeries: 1 },
			tooltipOptions: {
				formatTooltipY: (v) => "₹" + Number(v).toLocaleString("en-IN", { maximumFractionDigits: 0 }),
			},
		});
	}
}

File: instabiz/instabiz/page/ib_business_pulse/ib_business_pulse.py
import frappe
from frappe.utils import nowdate, getdate, get_first_day, get_last_day, add_months, flt


def get_context(context):
	context.no_cache = 1


@frappe.whitelist()
def get_pulse_data():
	today = getdate(nowdate())
	month_start = get_first_day(today)
	last_month_start = get_first_day(add_months(today, -1))
	last_month_end = get_last_day(add_months(today, -1))

	# ── Revenue ──────────────────────────────────────────────────────────────
	# TEST-ONLY BASIS CHANGE: Sales Order instead of Sales Invoice — billing
	# isn't live in ERP yet, so SI-based revenue always read 0/wrong. Revert to
	# Sales Invoice once invoicing goes live for real revenue-realization accounting.
	rev_mtd = flt(frappe.db.sql("""
		SELECT COALESCE(SUM(grand_total), 0)
		FROM `tabSales Order`
		WHERE docstatus=1
		AND transaction_date BETWEEN %s AND %s
	""", (month_start, today))[0][0])

	rev_last = flt(frappe.db.sql("""
		SELECT COALESCE(SUM(grand_total), 0)
		FROM `tabSales Order`
		WHERE docstatus=1
		AND transaction_date BETWEEN %s AND %s
	""", (last_month_start, last_month_end))[0][0])

	ar = flt(frappe.db.sql("""
		SELECT COALESCE(SUM(outstanding_amount), 0)
		FROM `tabSales Invoice`
		WHERE docstatus=1 AND outstanding_amount > 0
	""")[0][0])

	# Collection rate = collected / billed for the same 3-month window
	three_months_start = get_first_day(add_months(today, -3))
	total_billed = flt(frappe.db.sql("""
		SELECT COALESCE(SUM(grand_total), 0)
		FROM `tabSales Invoice`
		WHERE docstatus=1 AND is_return=0
		AND posting_date BETWEEN %s AND %s
	""", (three_months_start, today))[0][0])
	ar_3m = flt(frappe.db.sql("""
		SELECT COALESCE(SUM(outstanding_amount), 0)
		FROM `tabSales Invoice`
		WHERE docstatus=1 AND outstanding_amount > 0
		AND posting_date BETWEEN %s AND %s
	""", (three_months_start, today))[0][0])
	collected_3m = max(0.0, total_billed - ar_3m)
	collection_rate = max(0.0, min(100.0, round(collected_3m / total_billed * 100, 1))) if total_billed else 100.0
	rev_score = min(100, round((rev_mtd / rev_last * 80) if rev_last else 50, 0))

	# ── Sales Pipeline ───────────────────────────────────────────────────────
	open_leads = flt(frappe.db.sql("""
		SELECT COUNT(*) FROM `tabLead`
		WHERE status NOT IN ('Converted','Do Not Contact')
	""")[0][0])

	open_quotes = flt(frappe.db.sql("""
		SELECT COUNT(*) FROM `tabQuotation`
		WHERE docstatus=1 AND status NOT IN ('Ordered','Lost','Cancelled','Expired')
	""")[0][0])

	pipeline_score = min(100, int(open_leads * 5 + open_quotes * 10))

	# ── Inventory ────────────────────────────────────────────────────────────
	try:
		total_items = flt(frappe.db.sql(
			"SELECT COUNT(DISTINCT item_code) FROM `tabBin` WHERE actual_qty > 0"
		)[0][0])
		# Same definition as reorder_alert.py / ib_main_dashboard.py — actual_qty at
		# or below the item's reorder level, not just zero-stock-with-reservation.
		low_stock = flt(frappe.db.sql("""
			SELECT COUNT(DISTINCT b.item_code) FROM `tabBin` b
			INNER JOIN `tabItem Reorder` r
				ON r.parent = b.item_code AND r.warehouse = b.warehouse
			WHERE r.warehouse_reorder_level > 0
			  AND b.actual_qty <= r.warehouse_reorder_level
		""")[0][0])
		inv_score = max(0, round((1 - low_stock / total_items) * 100, 0)) if total_items else 80
	except Exception:
		total_items, low_stock, inv_score = 0, 0, 80

	# ── Procurement / Payables ───────────────────────────────────────────────
	try:
		open_po = flt(frappe.db.sql("""
			SELECT COUNT(*) FROM `tabPurchase Order`
			WHERE docstatus=1 AND status NOT IN ('Completed','Cancelled','Closed')
		""")[0][0])
		procurement_score = min(100, max(40, 100 - int(open_po * 3)))
	except Exception:
		open_po, procurement_score = 0, 80

	# ── HR ────────────────────────────────────────────────────────────────────
	try:
		total_emp = flt(frappe.db.sql(
			"SELECT COUNT(*) FROM `tabEmployee` WHERE status='Active'"
		)[0][0])
		present_today = flt(frappe.db.sql("""
			SELECT COUNT(DISTINCT employee) FROM (
				SELECT employee FROM `tabEmployee Checkin`
				WHERE DATE(time) = %s AND log_type = 'IN'
				UNION
				SELECT employee FROM `tabAttendance`
				WHERE attendance_date = %s AND status = 'Present' AND docstatus = 1
			) _combined
		""", (today, today))[0][0])
		hr_score = round((present_today / total_emp * 100), 0) if total_emp else 80
	except Exception:
		total_emp, present_today, hr_score = 0, 0, 80

	# ── Production ───────────────────────────────────────────────────────────
	try:
		wo_active = flt(frappe.db.sql("""
			SELECT COUNT(*) FROM `tabIB Work Order`
			WHERE status NOT IN ('Completed','Cancelled')
		""")[0][0])
		wo_completed = flt(frappe.db.sql("""
			SELECT COUNT(*) FROM `tabIB Work Order`
			WHERE status='Completed'
			AND COALESCE(completed_at, modified) >= %s
		""", (month_start,))[0][0])
		prod_total = wo_active + wo_completed
		prod_score = round((wo_completed / prod_total * 100), 0) if prod_total else 50
	except Exception:
		wo_active, wo_completed, prod_score = 0, 0, 80

	# ── 14-day revenue trend ──────────────────────────────────────────────────
	trend_14 = frappe.db.sql("""
		SELECT
			DATE_FORMAT(posting_date, '%%d %%b') as label,
			posting_date as dt,
			COALESCE(SUM(grand_total), 0) as amount
		FROM `tabSales Invoice`
		WHERE docstatus=1 AND is_return=0
		AND posting_date >= DATE_SUB(%s, INTERVAL 13 DAY)
		GROUP BY dt, label
		ORDER BY dt
	""", (today,), as_dict=True)

	# ── Overall health score (weighted avg) ───────────────────────────────────
	scores = {
		"Revenue":     float(rev_score),
		"Sales":       float(pipeline_score),
		"Inventory":   float(inv_score),
		"Procurement": float(procurement_score),
		"HR":          float(hr_score),
		"Production":  float(prod_score),
	}
	overall = round(sum(scores.values()) / len(scores), 0)

	return {
		"overall": overall,
		"scores": scores,
		"rev_mtd": rev_mtd,
		"rev_last": rev_last,
		"collection_rate": collection_rate,
		"ar": ar,
		"open_leads": int(open_leads),
		"open_quotes": int(open_quotes),
		"total_items": int(total_items),
		"low_stock": int(low_stock),
		"open_po": int(open_po),
		"total_emp": int(total_emp),
		"present_today": int(present_today),
		"wo_active": int(wo_active),
		"wo_completed": int(wo_completed),
		"trend_14": trend_14,
	}
